keps/sig-apps/3329-retriable-and-non-retriable-failures/README.md

```diff
@@ -0,0 +1,1267 @@
+failing due to the bugs in code of the executable, so that the computation resources
+can be saved.
+
+Occasionally, our executable fails, but it can be safely restarted with a good
+chance of succeeding the next time. In such known retriable situations our
+executable exits with a dedicated exit code in the 40-50 range. All remaining
+exit codes indicate a software bug and should result in an early job termination.
+
+The following Job configuration could be a good starting point to satisfy
+my needs:
+
+```yaml
+apiVersion: v1
+kind: Job
+spec:
+  template:
+    spec:
+      containers:
+      - name: job-container
+        image: job-image
+        command: ["./program"]
+  backoffLimit: 3
+  backoffPolicy:
+    rules:
+    - containerExitCodeNotIn: 40-50
+      action: Terminate
+```
+
+#### Story 2
+
+As a service provider that offers computational resources to researchers I would like to
+have a mechanism which terminates jobs for which pods are failing due to user errors,
+but allows infinite retries for infrastructure errors (pod terminations by kubelet,
+kube-scheduler, abnormal node termination, etc).
+I do not have knowledge or influence over the executable that researchers run,
+so I don't know beforehand which exit codes they might return.
+
+The following Job configuration could be a good starting point to satisfy
+my needs:
+
+```yaml
+apiVersion: v1
+kind: Job
+spec:
+  template:
+    spec:
+      containers:
+      - name: main-job-container
+        image: job-image
+        command: ["./program"]
+      - name: monitoring-job-container
+        image: job-monitoring
+        command: ["./monitoring"]
+  backoffLimit: 3
+  backoffPolicy:
+    rules:
+    - podStatusReasonIn:
+      - Preemption
+      - NodePressureEviction
+      action: Ignore
+```
+
+Note that, in this case the user supplies a list of pod's `status.reason` values
+which are interpreted as infrastructure failures. This approach is likely to
+require an iterative process to review and extend of the list. In order to
+avoid the process some users may prefer to interpret any non-empty `status.reason`
+as an infrastructure failure this can be achieved with the following syntax for
+`backoffPolicy`:
+
+```yaml
+  backoffPolicy:
+    rules:
+    - podStatusReasonNotIn:
+      - ''
+      action: Ignore
+```
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+#### Risk 1
+
+The list of available options for `reason` field will be evolving with new
+values being added and potentially some values becoming obsolete. This can make
+it difficult to maintain a valid list of reasons enumerated in the
+Job configuration.
+
+First, the users of the Job configuration will be able to mitigate it with the
```

Comment:
A mitigation should be on us to do, not on the user.

We can do 2 things:

- Document the existing reasons.
- Move all the reasons values to the `k8s.io/api/core/v1`. This would reduce typing mistakes and force an API reviewer to approve every new reason.